link-grammar/parse/fast-match.c

```diff
@@ -595,8 +595,8 @@ form_match_list(fast_matcher_t *ctxt, int w,
 	}
 
 #ifdef VERIFY_MATCH_LIST
-	static _Atomic(uint16_t) id = 0;
-	uint16_t lid = ++id; /* A local copy, for multi-threading support. */
+	static uint16_t id = 0;
+	uint16_t lid = ++id; /* A stable local copy, for multi-threading support. */
```

Comment:
This will clearly fail when multi-threading, as the increment `++id` can occur in two different threads (and so two threads get the same `lid`) which means that the assert at line 1470 of  `parse/count.c` might trigger.  I'm going to put this back, but then disable `VERIFY_MATCH_LIST` by default.

Reply:
Such a problem cannot happen, since each LG thread accesses only its own data structures!

Reply:
fixed in fac11ac24b33244f5a524f2b23cbc7c1165d7341

Reply:
Note that the lid is put in the element list that a `form_match_list()` returns, and since it is an auto variable, all the elements of the list are guaranteed to get the same lid. When  `form_match_list()` returns, the thread that has invoked it fetches the id from the first element and validates that the rest use the same id (to validate that the list has not been corrupted, as happens sometimes during development). So the code here doesn't care at all what lid values are used in other threads, and even not in the same thread (as long as lid is constant in a single invocation of form_match_list()`, which is guaranteed). They can be always the same, or any random ones.
The bug that got fixed than was due to an uninitialized id field on id==0.

Reply:
OK. I skimmed the code and saw this: 

https://github.com/opencog/link-grammar/blob/c09894035017c0cf1e4142b825ae4c23f3645175/link-grammar/parse/count.c#L1472

and (without over-analyzing it) made me think that assert might now trigger by accident. So I got nervous about it. So I put the Atomic back in, and undef'ed `VERIFY_MATCH_LIST` so it becomes invisible.

Reply:
Go ahead and remove the `_Atomic` if you are sure. I wasn't going to think too hard about it.